test/saithrift/tests/copp_tests.py

```diff
@@ -0,0 +1,549 @@
+        return
+
+    def copp_test(self, packet, count, send_intf, recv_intf):
+
+        start_time=datetime.datetime.now()
+        for i in xrange(count):
+            testutils.send_packet(self, send_intf, packet)
+
+        end_time=datetime.datetime.now()
+
+        total_rcv_pkt_cnt = 0
+        while True:
+            (rcv_device, rcv_port, rcv_pkt, pkt_time) = dp_poll(self, device_number=recv_intf[0], port_number=recv_intf[1], timeout=1)
+            if rcv_pkt is not None:
+                if match_exp_pkt(packet, rcv_pkt):
+                    total_rcv_pkt_cnt += 1
+            else:
+                break
+
+        time_delta = end_time - start_time
+        time_delta_ms = (time_delta.microseconds + time_delta.seconds * 10**6) / 10**3
+        if time_delta_ms == 0:
+            time_delta_ms = 1
+        
+        tx_pps = int(count/(float(time_delta_ms)/1000))
+        rx_pps = int(total_rcv_pkt_cnt/(float(time_delta_ms)/1000))
+        return total_rcv_pkt_cnt, time_delta, time_delta_ms, tx_pps, rx_pps
+
+    def contruct_packet(self, port_number):
+        raise NotImplemented
+
+    def check_constraints(self, total_rcv_pkt_cnt, time_delta_ms, rx_pps):
+        raise NotImplemented
+
+    def one_port_test(self, port_number, packet_count):
+
+        packet = self.contruct_packet(port_number)
+        total_rcv_pkt_cnt, time_delta, time_delta_ms, tx_pps, rx_pps = self.copp_test(packet, packet_count, (0, port_number), (1, port_number))
+        self.printStats(packet_count, total_rcv_pkt_cnt, time_delta, tx_pps, rx_pps)
+        self.check_constraints(total_rcv_pkt_cnt)
+        return
+
+    def run_suite(self):
+        raise NotImplemented
+
+    def printStats(self, pkt_send_count, total_rcv_pkt_cnt, time_delta, tx_pps, rx_pps):
+        if not(('verbose' in self.test_params) and (self.test_params['verbose'] == True)):
+            return
+        print 'test stats'
+        print 'Packet sent = %10d' % pkt_send_count
+        print 'Packet rcvd = %10d' % total_rcv_pkt_cnt
+        print 'Test time = %s' % str(time_delta)
+        print 'TX PPS = %d' % tx_pps
+        print 'RX PPS = %d' % rx_pps
+
+        return
+
+class NoPolicyTest(ControlPlaneBaseTest):
+    def __init__(self):
+        ControlPlaneBaseTest.__init__(self)
+
+    def run_suite(self):
+        port_to_test = self.test_params['test_port']
+        self.one_port_test(port_to_test, 1)
+
+    def check_constraints(self, total_rcv_pkt_cnt):
+        assert(total_rcv_pkt_cnt == 1)
+
+class PolicyTest(ControlPlaneBaseTest):
+    def __init__(self):
+        ControlPlaneBaseTest.__init__(self)
+
+    def check_constraints(self, total_rcv_pkt_cnt):
+        assert(self.POLICER_CIR*0.5 <= total_rcv_pkt_cnt <= self.POLICER_CIR*1.5)
+
+    def run_suite(self):
+        port_to_test = self.test_params['test_port']
+        self.one_port_test(port_to_test, self.PKT_TX_COUNT)
+
+# SONIC config contains policer  for ARP
+class ARPTest(PolicyTest):
```

Comment:
Each test should be self-contained, it should have follwoing setup

1. setup the sai configuration, trap id, trap group, policer, and etc. on
2. construct packet
3. send packet
4. verify packet
5. tear down the sai configuration.

For example in this ARP test, ARP related trap configuration should be in this class. Also, remove trap configuration should also be in this case.

With such separation, we can then test each test case-by-case.